### PR TITLE
Add AzureOpenAI support

### PR DESCRIPTION
**Pull request by sykuang**
I have created a new version to support AzureOpenAI as the previous pull request [#193](https://github.com/StanGirard/quivr/pull/193) was based on an outdated version, making it difficult to address the comments.
I have upgraded Langchain, OpenAI, and Guidance to support AzureOpenAI. However, in order to use the latest version of Langchain, we need to modify the ID type when using the add_documents function.

### CODE REVIEW

**Comment by StanGirard** [comment]
Thanks for the contribution. We need to create a migration script then. I'll take a look into it :)

File: backend/core/llm/openai_functions.py [hunk 1]
```diff
@@ -50,28 +55,53 @@ def __init__(
...
-        )  # pyright: ignore reportPrivateUsage=none
+        # Reference: langchain/chat_models/azure_openai.py
+        # We need use AzureChatOpenAI if we are using Azure Functions
+        if self.openai_api_type == "azure":
+            return AzureChatOpenAI(
+                openai_api_key=self.openai_api_key,
+                # to support Azure OpenAI Service custom endpoints
+                openai_api_base=self.openai_api_base,
+                openai_api_type=self.openai_api_type,
+                openai_api_version="2023-07-01-preview",
```
Comment (Zizico2) [inline backend/core/llm/openai_functions.py:86]: Shouldn't this be using the `OPENAI_API_VERSION` env var?

Comment (sykuang) [reply backend/core/llm/openai_functions.py]: Rebase the code and use the OPENAI_API_VERSION env var.

**Comment by igaspard** [comment]
This is the feature I really wanted 🤩

**Comment by Zizico2** [comment]
Just deployed it. Seems to be working.

File: .backend_env_azure.example [hunk 2]
```diff
@@ -0,0 +1,26 @@
+SUPABASE_URL=<change-me>
+SUPABASE_SERVICE_KEY=<change-me>
+PG_DATABASE_URL=notimplementedyet
+OPENAI_API_KEY=<change-me>
+OPENAI_API_TYPE=azure
+OPENAI_API_BASE=https://<Your-Endpoint>.openai.azure.com/
+OPENAI_GPT_DEPLOYMENT_ID=<change-me>
+OPENAI_EMBEDDING_DEPLOYMENT_ID=<change-me>
+OPENAI_API_VERSION="2023-07-01-preview"
```
Comment (Zizico2) [inline .backend_env_azure.example:9]: > OPENAI_GPT_DEPLOYMENT_ID=<change-me>
> OPENAI_EMBEDDING_DEPLOYMENT_ID=<change-me>
> OPENAI_API_VERSION="2023-07-01-preview"

These 3 are Azure specific. Would it make sense to prefix them with `AZURE_`? 

AutoGPT uses these names:

> azure_api_type: azure
> azure_api_base: your-base-url-for-azure
> azure_api_version: api-version-for-azure
> azure_model_map:
>     fast_llm_deployment_id: gpt35-deployment-id-for-azure
>     smart_llm_deployment_id: gpt4-deployment-id-for-azure
>     embedding_model_deployment_id: embedding-deployment-id-for-azure

Comment (sykuang) [reply .backend_env_azure.example]: Make sense. Rename those environments

**Review by Zizico2 — APPROVED** [review]

**Comment by krrishdholakia** [comment]
I assume this is covered by the litellm integration

**Comment by krrishdholakia** [comment]
https://docs.litellm.ai/docs/providers/azure

**Pull request closed without merging**